File: src/lib/calculations.ts
import { BillingPortfolioBreakdown, BillingPortfolioCode, Site, SiteWithCalculations, RateTier, PortfolioSummary, SitePricingBreakdown } from '@/types';

// Default rate tiers matching the spreadsheet
export const DEFAULT_RATE_TIERS: RateTier[] = [
  { id: '1', contractId: 'DEFAULT', tierName: '<20MW', minCapacityMW: 0, maxCapacityMW: 20, ratePerKwp: 2.0 },
  { id: '2', contractId: 'DEFAULT', tierName: '20-30MW', minCapacityMW: 20, maxCapacityMW: 30, ratePerKwp: 1.8 },
  { id: '3', contractId: 'DEFAULT', tierName: '30-40MW', minCapacityMW: 30, maxCapacityMW: 40, ratePerKwp: 1.7 },
];

export function calculateSiteFixedCosts(site: Site): number {
  return site.pmCost + site.cctvCost + site.cleaningCost + (site.additionalCostAnnual || 0);
}

export function calculatePortfolioCost(systemSizeKwp: number, ratePerKwp: number): number {
  return systemSizeKwp * ratePerKwp;
}

export function calculateFixedFee(siteFixedCosts: number, portfolioCost: number): number {
  return siteFixedCosts + portfolioCost;
}

export function calculateFeePerKwp(fixedFee: number, systemSizeKwp: number, isContracted: boolean): number {
  if (!isContracted || systemSizeKwp === 0) return 0;
  return fixedFee / systemSizeKwp;
}

export function calculateMonthlyFee(fixedFee: number): number {
  return fixedFee / 12;
}

export function billingPortfolioLabel(portfolio: BillingPortfolioCode): string {
  return portfolio === 'EDEN' ? 'Eden' : 'Core';
}

export function determinePortfolioTier(totalCapacityMW: number, tiers: RateTier[] = DEFAULT_RATE_TIERS): RateTier {
  // Find the appropriate tier based on total capacity
  for (const tier of tiers) {
    if (tier.maxCapacityMW === null || totalCapacityMW < tier.maxCapacityMW) {
      return tier;
    }
  }
  return tiers[tiers.length - 1];
}

function tierByName(tiers: RateTier[], name: string, fallbackIndex: number): RateTier {
  return tiers.find(t => t.tierName === name) || tiers[fallbackIndex] || tiers[0];
}

export function isContractedStatus(status: string): boolean {
  return ['Contracted', 'CONTRACTED', 'Yes', 'YES', 'Active', 'ACTIVE'].includes(status);
}

export function siteBillingStartDate(site: { actualPacDate?: string | null; onboardDate: string | null }): string | null {
  return site.onboardDate || site.actualPacDate || null;
}

export function isSiteVisibleByMonthEnd(site: { actualPacDate?: string | null; onboardDate: string | null }, monthEnd?: Date | null): boolean {
  const startDate = siteBillingStartDate(site);
  if (!monthEnd || !startDate) return true;
  return new Date(`${startDate.slice(0, 10)}T00:00:00.000Z`) <= monthEnd;
}

export function contractedCapacityForTier(
  sites: Array<Pick<Site, 'contractStatus' | 'systemSizeKwp' | 'onboardDate'> & { actualPacDate?: string | null }>,
  monthEnd?: Date | null
): number {
  return sites
    .filter((site) => isContractedStatus(site.contractStatus) && isSiteVisibleByMonthEnd(site, monthEnd))
    .reduce((sum, site) => sum + site.systemSizeKwp, 0);
}

export function currentPortfolioTier(
  sites: Array<Pick<Site, 'contractStatus' | 'systemSizeKwp' | 'onboardDate'> & { actualPacDate?: string | null }>,
  tiers: RateTier[] = DEFAULT_RATE_TIERS,
  monthEnd?: Date | null
): RateTier {
  return determinePortfolioTier(contractedCapacityForTier(sites, monthEnd) / 1000, tiers);
}

export function calculateAnnualFeeForTier(site: Site, tier: RateTier): number {
  return calculateAnnualFeeForTierForMonth(site, tier);
}

function isMonthValue(value: string | null | undefined): value is string {
  return /^\d{4}-(0[1-9]|1[0-2])$/.test(value || '');
}

export function isAdditionalMonthlyCostActive(site: Site, month?: string | null): boolean {
  if (!site.additionalCostMonthly) return false;
  if (!month) return true;

  const normalizedMonth = isMonthValue(month) ? month : null;
  if (!normalizedMonth) return true;

  const startMonth = isMonthValue(site.additionalCostMonthlyStartMonth) ? site.additionalCostMonthlyStartMonth : null;
  const endMonth = isMonthValue(site.additionalCostMonthlyEndMonth) ? site.additionalCostMonthlyEndMonth : null;

  if (startMonth && normalizedMonth < startMonth) return false;
  if (endMonth && normalizedMonth > endMonth) return false;
  return true;
}

export function calculateAnnualFeeForTierForMonth(site: Site, tier: RateTier, month?: string | null): number {
  if (!isContractedStatus(site.contractStatus)) return 0;
  return calculateFixedFee(calculateSiteFixedCosts(site), calculatePortfolioCost(site.systemSizeKwp, tier.ratePerKwp)) + calculateAdditionalMonthlyAnnual(site, month);
}

function calculateAdditionalMonthlyAnnual(site: Site, month?: string | null): number {
  return isAdditionalMonthlyCostActive(site, month) ? (site.additionalCostMonthly || 0) * 12 : 0;
}

export function buildSitePricingBreakdown(
  site: Site,
  appliedTier: RateTier,
  contractedCapacityKwpForTier: number,
  month?: string | null
): SitePricingBreakdown {
  const isBillable = isContractedStatus(site.contractStatus);
  const siteFixedCostsAnnual = calculateSiteFixedCosts(site);
  const billablePortfolioCostAnnual = calculatePortfolioCost(site.systemSizeKwp, appliedTier.ratePerKwp);
  const billableAdditionalMonthlyAnnual = calculateAdditionalMonthlyAnnual(site, month);
  const portfolioCostAnnual = isBillable ? billablePortfolioCostAnnual : 0;
  const additionalMonthlyAnnual = isBillable ? billableAdditionalMonthlyAnnual : 0;
  const annualFee = isBillable ? calculateAnnualFeeForTierForMonth(site, appliedTier, month) : 0;
  const monthlyFee = isBillable ? calculateMonthlyFee(annualFee) : 0;
  const nonBillableNote = isBillable ? null : 'Excluded because site is not contracted';

  return {
    isBillable,
    reviewStatus: isBillable ? 'VERIFIED' : 'NEEDS_REVIEW',
    reviewReason: isBillable
      ? 'Contracted site priced using active portfolio tier'
      : `Not billable while contract status is ${site.contractStatus}`,
    appliedTierName: appliedTier.tierName,
    appliedTierRatePerKwp: appliedTier.ratePerKwp,
    contractedCapacityKwpForTier,
    siteFixedCostsAnnual,
    portfolioCostAnnual,
    additionalMonthlyAnnual,
    annualFee,
    monthlyFee,
    formula: '(site fixed costs + capacity x rate + active monthly extras x 12) / 12',
    lines: [
      {
        label: 'PM cost',
        calculation: 'Imported fixed cost',
        annualValue: site.pmCost,
      },
      {
        label: 'CCTV cost',
        calculation: 'Imported fixed cost',
        annualValue: site.cctvCost,
      },
      {
        label: 'Cleaning cost',
        calculation: 'Imported fixed cost',
        annualValue: site.cleaningCost,
      },
      {
        label: 'Additional annual cost',
        calculation: 'Imported fixed cost',
        annualValue: site.additionalCostAnnual || 0,
        note: site.additionalCostAnnualComment,
      },
      {
        label: 'Portfolio tariff',
        calculation: `${formatNumber(site.systemSizeKwp, 0)} kWp x ${formatCurrency(appliedTier.ratePerKwp)}/kWp`,
        annualValue: portfolioCostAnnual,
        note: nonBillableNote,
      },
      {
        label: 'Additional monthly cost',
        calculation: `${formatCurrency(site.additionalCostMonthly || 0)} x 12`,
        annualValue: additionalMonthlyAnnual,
        note: isBillable
          ? site.additionalCostMonthlyComment
          : nonBillableNote,
      },
      {
        label: 'Annual fee',
        calculation: 'Sum above',
        annualValue: annualFee,
        note: nonBillableNote,
      },
      {
        label: 'Monthly fee',
        calculation: `${formatCurrency(annualFee)} / 12`,
        annualValue: monthlyFee,
        note: nonBillableNote,
      },
    ],
  };
}

export function calculateSiteWithAllTiers(
  site: Site,
  tiers: RateTier[] = DEFAULT_RATE_TIERS,
  appliedTier?: RateTier,
  month?: string | null,
  contractedCapacityKwpForTier?: number
): SiteWithCalculations {
  const siteFixedCosts = calculateSiteFixedCosts(site);
  const isContracted = isContractedStatus(site.contractStatus);
  
  // Calculate for each tier
  const tier20MW = tierByName(tiers, '<20MW', 0);
  const tier30MW = tierByName(tiers, '20-30MW', 1);
  const tier40MW = tierByName(tiers, '30-40MW', 2);
  
  const portfolioCost_20MW = calculatePortfolioCost(site.systemSizeKwp, tier20MW.ratePerKwp);
  const portfolioCost_30MW = calculatePortfolioCost(site.systemSizeKwp, tier30MW.ratePerKwp);
  const portfolioCost_40MW = calculatePortfolioCost(site.systemSizeKwp, tier40MW.ratePerKwp);
  
  const additionalMonthlyAnnual = calculateAdditionalMonthlyAnnual(site, month);
  const fixedFee_20MW = calculateFixedFee(siteFixedCosts, portfolioCost_20MW) + additionalMonthlyAnnual;
  const fixedFee_30MW = calculateFixedFee(siteFixedCosts, portfolioCost_30MW) + additionalMonthlyAnnual;
  const fixedFee_40MW = calculateFixedFee(siteFixedCosts, portfolioCost_40MW) + additionalMonthlyAnnual;
  
  const feePerKwp_20MW = calculateFeePerKwp(fixedFee_20MW, site.systemSizeKwp, isContracted);
  const feePerKwp_30MW = calculateFeePerKwp(fixedFee_30MW, site.systemSizeKwp, isContracted);
  const feePerKwp_40MW = calculateFeePerKwp(fixedFee_40MW, site.systemSizeKwp, isContracted);
  
  // Monthly fee based on the active portfolio tier, which is derived from contracted capacity only.
  const selectedTier = appliedTier || tier20MW;
  const monthlyFee = isContracted ? calculateMonthlyFee(calculateAnnualFeeForTierForMonth(site, selectedTier, month)) : 0;
  
  return {
    ...site,
    siteFixedCosts,
    portfolioCost_20MW,
    portfolioCost_30MW,
    portfolioCost_40MW,
    fixedFee_20MW,
    fixedFee_30MW,
    fixedFee_40MW,
    feePerKwp_20MW,
    feePerKwp_30MW,
    feePerKwp_40MW,
    monthlyFee,
    pricingBreakdown: buildSitePricingBreakdown(site, selectedTier, contractedCapacityKwpForTier ?? 0, month),
  };
}

function emptyBillingPortfolioBreakdown(billingPortfolio: BillingPortfolioCode): BillingPortfolioBreakdown {
  return {
    billingPortfolio,
    billingPortfolioLabel: billingPortfolioLabel(billingPortfolio),
    siteCount: 0,
    contractedSiteCount: 0,
    pendingSiteCount: 0,
    totalCapacityKwp: 0,
    contractedCapacityKwp: 0,
    siteFixedCostsAnnual: 0,
    variableCostAnnual: 0,
    annualFee: 0,
    monthlyFee: 0,
    correctiveDaysAllowed: 0,
  };
}

function finalizeBillingPortfolioBreakdown(row: BillingPortfolioBreakdown): BillingPortfolioBreakdown {
  return {
    ...row,
    correctiveDaysAllowed: Math.floor(row.contractedCapacityKwp / 1000 / 12),
  };
}

export function buildBillingPortfolioBreakdowns(sites: Site[], tier: RateTier): BillingPortfolioBreakdown[] {
  const breakdowns = new Map<BillingPortfolioCode, BillingPortfolioBreakdown>([
    ['CORE', emptyBillingPortfolioBreakdown('CORE')],
    ['EDEN', emptyBillingPortfolioBreakdown('EDEN')],
  ]);

  for (const site of sites) {
    const billingPortfolio = site.billingPortfolio === 'EDEN' ? 'EDEN' : 'CORE';
    const row = breakdowns.get(billingPortfolio)!;
    row.siteCount += 1;
    row.totalCapacityKwp += site.systemSizeKwp;

    if (isContractedStatus(site.contractStatus)) {
      const siteFixedCosts = calculateSiteFixedCosts(site);
      const annualFee = calculateAnnualFeeForTier(site, tier);
      row.contractedSiteCount += 1;
      row.contractedCapacityKwp += site.systemSizeKwp;
      row.siteFixedCostsAnnual += siteFixedCosts;
      row.variableCostAnnual += Math.max(annualFee - siteFixedCosts, 0);
      row.annualFee += annualFee;
      row.monthlyFee += annualFee / 12;
    } else {
      row.pendingSiteCount += 1;
    }
  }

  return Array.from(breakdowns.values()).map(finalizeBillingPortfolioBreakdown);
}

export function calculatePortfolioSummary(sites: Site[], tiers: RateTier[] = DEFAULT_RATE_TIERS): PortfolioSummary {
  const contractedSites = sites.filter(s => isContractedStatus(s.contractStatus));
  const totalCapacityKwp = sites.reduce((sum, s) => sum + s.systemSizeKwp, 0);
  const contractedCapacityKwp = contractedSites.reduce((sum, s) => sum + s.systemSizeKwp, 0);
  
  const currentTier = determinePortfolioTier(contractedCapacityKwp / 1000, tiers);
  
  // Calculate total monthly fee
  const sitesWithCalcs = contractedSites.map(s => calculateSiteWithAllTiers(s, tiers, currentTier));
  const totalMonthlyFee = sitesWithCalcs.reduce((sum, s) => sum + s.monthlyFee, 0);
  
  // Corrective days calculation
  const correctiveDaysAllowed = Math.floor(contractedCapacityKwp / 1000 / 12);
  
  // Sites by SPV
  const sitesBySpv: Record<string, number> = {};
  sites.forEach(site => {
    const spv = site.spvCode || 'Unassigned';
    sitesBySpv[spv] = (sitesBySpv[spv] || 0) + 1;
  });
  
  return {
    totalSites: sites.length,
    contractedSites: contractedSites.length,
    totalCapacityKwp,
    contractedCapacityKwp,
    currentTier: currentTier.tierName,
    totalMonthlyFee,
    correctiveDaysAllowed,
    sitesBySpv,
    portfolioBreakdowns: buildBillingPortfolioBreakdowns(sites, currentTier),
  };
}

export function formatCurrency(value: number): string {
  return new Intl.NumberFormat('en-GB', {
    style: 'currency',
    currency: 'GBP',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(value);
}

export function formatNumber(value: number, decimals: number = 2): string {
  return new Intl.NumberFormat('en-GB', {
    minimumFractionDigits: decimals,
    maximumFractionDigits: decimals,
  }).format(value);
}

File: streamlit_app/calculations.py
"""
Fee calculation logic for Clearsol O&M Portfolio Tracker.
Replicates the spreadsheet formulas exactly.
"""

from typing import Optional
import db


# ============ Default Rate Tiers ============

DEFAULT_RATE_TIERS = [
    {'id': '1', 'tier_name': '<20MW', 'min_capacity_mw': 0, 'max_capacity_mw': 20, 'rate_per_kwp': 2.0},
    {'id': '2', 'tier_name': '20-30MW', 'min_capacity_mw': 20, 'max_capacity_mw': 30, 'rate_per_kwp': 1.8},
    {'id': '3', 'tier_name': '30-40MW', 'min_capacity_mw': 30, 'max_capacity_mw': 40, 'rate_per_kwp': 1.7},
]


# ============ Core Calculation Functions ============

def calculate_site_fixed_costs(pm_cost: float, cctv_cost: float, cleaning_cost: float) -> float:
    """
    Calculate site fixed costs.
    Site Fixed Costs = PM Cost + CCTV Cost + Cleaning Cost
    """
    return pm_cost + cctv_cost + cleaning_cost


def calculate_portfolio_cost(system_size_kwp: float, rate_per_kwp: float) -> float:
    """
    Calculate portfolio cost for a site.
    Portfolio Cost = System Size (kWp) × Rate per kWp (tier-based)
    """
    return system_size_kwp * rate_per_kwp


def calculate_fixed_fee(site_fixed_costs: float, portfolio_cost: float) -> float:
    """
    Calculate fixed fee.
    Fixed Fee = Site Fixed Costs + Portfolio Cost
    """
    return site_fixed_costs + portfolio_cost


def calculate_fee_per_kwp(fixed_fee: float, system_size_kwp: float, is_contracted: bool) -> float:
    """
    Calculate fee per kWp.
    Fee per kWp = Fixed Fee / System Size (only if contracted)
    """
    if not is_contracted or system_size_kwp == 0:
        return 0
    return fixed_fee / system_size_kwp


def calculate_monthly_fee(fixed_fee: float) -> float:
    """
    Calculate monthly fee.
    Monthly Fee = Fixed Fee / 12
    """
    return fixed_fee / 12


def determine_portfolio_tier(total_capacity_mw: float, tiers: Optional[list[dict]] = None) -> dict:
    """
    Determine the appropriate rate tier based on total portfolio capacity.
    """
    if tiers is None:
        tiers = db.get_rate_tiers() or DEFAULT_RATE_TIERS
    
    for tier in tiers:
        max_cap = tier.get('max_capacity_mw')
        if max_cap is None or total_capacity_mw < max_cap:
            return tier
    
    return tiers[-1] if tiers else DEFAULT_RATE_TIERS[-1]


def calculate_corrective_days(contracted_capacity_kwp: float) -> float:
    """
    Calculate corrective days allowed.
    Corrective Days = Capacity / 1000 / 12
    """
    return round((contracted_capacity_kwp / 1000 / 12) * 10) / 10


# ============ Site Calculation Functions ============

def calculate_site_with_all_tiers(site: dict, tiers: Optional[list[dict]] = None) -> dict:
    """
    Calculate all fee metrics for a site across all tiers.
    Returns the site dict with additional calculated fields.
    """
    if tiers is None:
        tiers = db.get_rate_tiers() or DEFAULT_RATE_TIERS
    
    # Get tier rates
    tier_20mw = next((t for t in tiers if t['tier_name'] == '<20MW'), tiers[0] if tiers else DEFAULT_RATE_TIERS[0])
    tier_30mw = next((t for t in tiers if t['tier_name'] == '20-30MW'), tiers[1] if len(tiers) > 1 else DEFAULT_RATE_TIERS[1])
    tier_40mw = next((t for t in tiers if t['tier_name'] == '30-40MW'), tiers[2] if len(tiers) > 2 else DEFAULT_RATE_TIERS[2])
    
    # Calculate site fixed costs
    site_fixed_costs = calculate_site_fixed_costs(
        site.get('pm_cost', 0),
        site.get('cctv_cost', 0),
        site.get('cleaning_cost', 0)
    )
    
    system_size = site.get('system_size_kwp', 0)
    is_contracted = site.get('contract_status') == 'Yes'
    
    # Calculate portfolio costs for each tier
    portfolio_cost_20mw = calculate_portfolio_cost(system_size, tier_20mw['rate_per_kwp'])
    portfolio_cost_30mw = calculate_portfolio_cost(system_size, tier_30mw['rate_per_kwp'])
    portfolio_cost_40mw = calculate_portfolio_cost(system_size, tier_40mw['rate_per_kwp'])
    
    # Calculate fixed fees for each tier
    fixed_fee_20mw = calculate_fixed_fee(site_fixed_costs, portfolio_cost_20mw)
    fixed_fee_30mw = calculate_fixed_fee(site_fixed_costs, portfolio_cost_30mw)
    fixed_fee_40mw = calculate_fixed_fee(site_fixed_costs, portfolio_cost_40mw)
    
    # Calculate fee per kWp for each tier
    fee_per_kwp_20mw = calculate_fee_per_kwp(fixed_fee_20mw, system_size, is_contracted)
    fee_per_kwp_30mw = calculate_fee_per_kwp(fixed_fee_30mw, system_size, is_contracted)
    fee_per_kwp_40mw = calculate_fee_per_kwp(fixed_fee_40mw, system_size, is_contracted)
    
    # Monthly fee based on <20MW tier (default)
    monthly_fee = calculate_monthly_fee(fixed_fee_20mw) if is_contracted else 0
    
    # Return site with calculations
    return {
        **site,
        'site_fixed_costs': site_fixed_costs,
        'portfolio_cost_20mw': portfolio_cost_20mw,
        'portfolio_cost_30mw': portfolio_cost_30mw,
        'portfolio_cost_40mw': portfolio_cost_40mw,
        'fixed_fee_20mw': fixed_fee_20mw,
        'fixed_fee_30mw': fixed_fee_30mw,
        'fixed_fee_40mw': fixed_fee_40mw,
        'fee_per_kwp_20mw': fee_per_kwp_20mw,
        'fee_per_kwp_30mw': fee_per_kwp_30mw,
        'fee_per_kwp_40mw': fee_per_kwp_40mw,
        'monthly_fee': monthly_fee,
    }


def calculate_portfolio_summary(sites: list[dict]) -> dict:
    """
    Calculate portfolio-level summary statistics.
    """
    contracted_sites = [s for s in sites if s.get('contract_status') == 'Yes']
    
    total_capacity_kwp = sum(s.get('system_size_kwp', 0) for s in sites)
    contracted_capacity_kwp = sum(s.get('system_size_kwp', 0) for s in contracted_sites)
    
    # Determine current tier
    current_tier = determine_portfolio_tier(contracted_capacity_kwp / 1000)
    
    # Calculate total monthly fee
    sites_with_calcs = [calculate_site_with_all_tiers(s) for s in contracted_sites]
    total_monthly_fee = sum(s.get('monthly_fee', 0) for s in sites_with_calcs)
    
    # Corrective days calculation
    corrective_days_allowed = calculate_corrective_days(contracted_capacity_kwp)
    
    # Sites by SPV
    sites_by_spv = {}
    for site in sites:
        spv = site.get('spv_code') or 'Unassigned'
        sites_by_spv[spv] = sites_by_spv.get(spv, 0) + 1
    
    return {
        'total_sites': len(sites),
        'contracted_sites': len(contracted_sites),
        'total_capacity_kwp': total_capacity_kwp,
        'contracted_capacity_kwp': contracted_capacity_kwp,
        'current_tier': current_tier.get('tier_name', 'N/A'),
        'total_monthly_fee': total_monthly_fee,
        'corrective_days_allowed': corrective_days_allowed,
        'sites_by_spv': sites_by_spv,
    }


# ============ Formatting Functions ============

def format_currency(value: float) -> str:
    """Format a number as GBP currency."""
    return f"£{value:,.2f}"


def format_number(value: float, decimals: int = 2) -> str:
    """Format a number with specified decimal places."""
    return f"{value:,.{decimals}f}"
